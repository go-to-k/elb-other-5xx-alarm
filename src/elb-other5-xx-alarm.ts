import { Duration } from 'aws-cdk-lib';
import {
  Alarm,
  AlarmRule,
  AlarmState,
  ComparisonOperator,
  CompositeAlarm,
  IAlarmAction,
  Metric,
  TreatMissingData,
} from 'aws-cdk-lib/aws-cloudwatch';
import { Construct } from 'constructs';

export interface ELBOther5XXAlarmProps {
  /**
   * Full name of the target LoadBalancer.
   */
  readonly loadBalancerFullName: string;

  /**
   * Alarm actions using in this alarm.
   */
  readonly alarmActions: IAlarmAction[];

  /**
   * The period over which the specified statistic is applied.
   *
   * @default Duration.minutes(5)
   */
  readonly period?: Duration;

  /**
   * Name of the alarm.
   *
   * @default Automatically generated name
   */
  readonly alarmName?: string;

  /**
   * Description for the alarm.
   *
   * @default No description
   */
  readonly alarmDescription?: string;

  /**
   * The value against which the specified statistic is compared.
   *
   */
  readonly threshold: number;

  /**
   * The number of periods over which data is compared to the specified threshold.
   *
   */
  readonly evaluationPeriods: number;

  /**
   * The number of datapoints that must be breaching to trigger the alarm.
   *
   * This is used only if you are setting an "M
   * out of N" alarm. In that case, this value is the M. For more information, see Evaluating an Alarm in the Amazon
   * CloudWatch User Guide.
   *
   * @default ``evaluationPeriods``
   * @see https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/AlarmThatSendsEmail.html#alarm-evaluation
   */
  readonly datapointsToAlarm?: number;
}

export class ELBOther5XXAlarm extends Construct {
  /**
   * The composite alarm resource for HTTPCode_ELB_5XX_Count excluding 500, 502, 503, and 504.
   */
  public readonly elbOther5XXAlarm: CompositeAlarm;

  constructor(scope: Construct, id: string, props: ELBOther5XXAlarmProps) {
    super(scope, id);

    const httpCodeELB5XXCountAlarm = new Alarm(this, 'HTTPCodeELB5XXCountAlarm', {
      metric: new Metric({
        namespace: 'AWS/ApplicationELB',
        metricName: 'HTTPCode_ELB_5XX_Count',
        dimensionsMap: { LoadBalancer: props.loadBalancerFullName },
        statistic: 'Sum',
        period: props.period,
      }),
      actionsEnabled: false,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      threshold: props.threshold,
      treatMissingData: TreatMissingData.NOT_BREACHING,
      evaluationPeriods: props.evaluationPeriods,
      datapointsToAlarm: props.datapointsToAlarm,
    });

    const httpCodeELB500CountAlarm = new Alarm(this, 'HTTPCodeELB500CountAlarm', {
      metric: new Metric({
        namespace: 'AWS/ApplicationELB',
        metricName: 'HTTPCode_ELB_500_Count',
        dimensionsMap: { LoadBalancer: props.loadBalancerFullName },
        statistic: 'Sum',
        period: props.period,
      }),
      actionsEnabled: false,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      threshold: props.threshold,
      treatMissingData: TreatMissingData.NOT_BREACHING,
      evaluationPeriods: props.evaluationPeriods,
      datapointsToAlarm: props.datapointsToAlarm,
    });

    const httpCodeELB502CountAlarm = new Alarm(this, 'HTTPCodeELB502CountAlarm', {
      metric: new Metric({
        namespace: 'AWS/ApplicationELB',
        metricName: 'HTTPCode_ELB_502_Count',
        dimensionsMap: { LoadBalancer: props.loadBalancerFullName },
        statistic: 'Sum',
        period: props.period,
      }),
      actionsEnabled: false,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      threshold: props.threshold,
      treatMissingData: TreatMissingData.NOT_BREACHING,
      evaluationPeriods: props.evaluationPeriods,
      datapointsToAlarm: props.datapointsToAlarm,
    });

    const httpCodeELB503CountAlarm = new Alarm(this, 'HTTPCodeELB503CountAlarm', {
      metric: new Metric({
        namespace: 'AWS/ApplicationELB',
        metricName: 'HTTPCode_ELB_503_Count',
        dimensionsMap: { LoadBalancer: props.loadBalancerFullName },
        statistic: 'Sum',
        period: props.period,
      }),
      actionsEnabled: false,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      threshold: props.threshold,
      treatMissingData: TreatMissingData.NOT_BREACHING,
      evaluationPeriods: props.evaluationPeriods,
      datapointsToAlarm: props.datapointsToAlarm,
    });

    const httpCodeELB504CountAlarm = new Alarm(this, 'HTTPCodeELB504CountAlarm', {
      metric: new Metric({
        namespace: 'AWS/ApplicationELB',
        metricName: 'HTTPCode_ELB_504_Count',
        dimensionsMap: { LoadBalancer: props.loadBalancerFullName },
        statistic: 'Sum',
        period: props.period,
      }),
      actionsEnabled: false,
      comparisonOperator: ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      threshold: props.threshold,
      treatMissingData: TreatMissingData.NOT_BREACHING,
      evaluationPeriods: props.evaluationPeriods,
      datapointsToAlarm: props.datapointsToAlarm,
    });

    const elbConcreteStatusSuppressorAlarm = new CompositeAlarm(
      this,
      'ELBConcreteStatusSuppressorAlarm',
      {
        actionsEnabled: false,
        alarmRule: AlarmRule.anyOf(
          AlarmRule.fromAlarm(httpCodeELB500CountAlarm, AlarmState.ALARM),
          AlarmRule.fromAlarm(httpCodeELB502CountAlarm, AlarmState.ALARM),
          AlarmRule.fromAlarm(httpCodeELB503CountAlarm, AlarmState.ALARM),
          AlarmRule.fromAlarm(httpCodeELB504CountAlarm, AlarmState.ALARM),
        ),
      },
    );

    this.elbOther5XXAlarm = new CompositeAlarm(this, 'Resource', {
      compositeAlarmName: props.alarmName,
      alarmDescription: props.alarmDescription,
      actionsEnabled: true,
      alarmRule: AlarmRule.allOf(
        AlarmRule.fromAlarm(httpCodeELB5XXCountAlarm, AlarmState.ALARM),
        AlarmRule.not(AlarmRule.fromAlarm(elbConcreteStatusSuppressorAlarm, AlarmState.ALARM)),
      ),
      actionsSuppressor: elbConcreteStatusSuppressorAlarm,
      actionsSuppressorExtensionPeriod: Duration.seconds(60),
      actionsSuppressorWaitPeriod: Duration.seconds(60),
    });
    this.elbOther5XXAlarm.addAlarmAction(...props.alarmActions);
  }
}
