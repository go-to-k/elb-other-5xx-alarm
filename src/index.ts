export * from './elb-other5-xx-alarm';
