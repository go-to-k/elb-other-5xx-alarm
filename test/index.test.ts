import { App, Duration, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { ELBOther5XXAlarm } from '../src';

const ALARM_NAME = 'test-alarm';

const getTemplate = (): Template => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const topic = new Topic(stack, 'Topic', {});
  const alarmActions = [new SnsAction(topic)];

  new ELBOther5XXAlarm(stack, 'ELBOther5XXAlarm', {
    loadBalancerFullName: 'default',
    alarmName: ALARM_NAME,
    alarmActions: alarmActions,
    period: Duration.seconds(60),
    threshold: 1,
    evaluationPeriods: 1,
  });
  const template = Template.fromStack(stack);

  return template;
};

describe('Fine-grained Assertions Tests', () => {
  const template = getTemplate();

  test('Alarm created', () => {
    template.resourceCountIs('AWS::CloudWatch::Alarm', 5);
  });

  test('CompositeAlarm created', () => {
    template.resourceCountIs('AWS::CloudWatch::CompositeAlarm', 2);
  });

  test('CompositeAlarm created', () => {
    template.hasResourceProperties('AWS::CloudWatch::CompositeAlarm', {
      AlarmName: ALARM_NAME,
      ActionsEnabled: true,
    });
  });
});
