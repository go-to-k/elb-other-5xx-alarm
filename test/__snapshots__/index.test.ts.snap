// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fine-grained Assertions Tests Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ELBOther5XXAlarm97BD7EA5": {
      "Properties": {
        "ActionsEnabled": true,
        "ActionsSuppressor": {
          "Fn::GetAtt": [
            "ELBOther5XXAlarmELBConcreteStatusSuppressorAlarm74C416F3",
            "Arn",
          ],
        },
        "ActionsSuppressorExtensionPeriod": 60,
        "ActionsSuppressorWaitPeriod": 60,
        "AlarmActions": [
          {
            "Ref": "TopicBFC7AF6E",
          },
        ],
        "AlarmName": "test-alarm",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ELBOther5XXAlarmHTTPCodeELB5XXCountAlarmC0D9FEA7",
                  "Arn",
                ],
              },
              "") AND (NOT (ALARM("",
              {
                "Fn::GetAtt": [
                  "ELBOther5XXAlarmELBConcreteStatusSuppressorAlarm74C416F3",
                  "Arn",
                ],
              },
              ""))))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "ELBOther5XXAlarmELBConcreteStatusSuppressorAlarm74C416F3": {
      "Properties": {
        "ActionsEnabled": false,
        "AlarmName": "TestStackELBOther5XXAlarmELBConcreteStatusSuppressorAlarm875CB7FD",
        "AlarmRule": {
          "Fn::Join": [
            "",
            [
              "(ALARM("",
              {
                "Fn::GetAtt": [
                  "ELBOther5XXAlarmHTTPCodeELB500CountAlarmE024342E",
                  "Arn",
                ],
              },
              "") OR ALARM("",
              {
                "Fn::GetAtt": [
                  "ELBOther5XXAlarmHTTPCodeELB502CountAlarmFEB9F989",
                  "Arn",
                ],
              },
              "") OR ALARM("",
              {
                "Fn::GetAtt": [
                  "ELBOther5XXAlarmHTTPCodeELB503CountAlarm9245609E",
                  "Arn",
                ],
              },
              "") OR ALARM("",
              {
                "Fn::GetAtt": [
                  "ELBOther5XXAlarmHTTPCodeELB504CountAlarm3CB4C2A2",
                  "Arn",
                ],
              },
              ""))",
            ],
          ],
        },
      },
      "Type": "AWS::CloudWatch::CompositeAlarm",
    },
    "ELBOther5XXAlarmHTTPCodeELB500CountAlarmE024342E": {
      "Properties": {
        "ActionsEnabled": false,
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "LoadBalancer",
            "Value": "default",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "HTTPCode_ELB_500_Count",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ELBOther5XXAlarmHTTPCodeELB502CountAlarmFEB9F989": {
      "Properties": {
        "ActionsEnabled": false,
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "LoadBalancer",
            "Value": "default",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "HTTPCode_ELB_502_Count",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ELBOther5XXAlarmHTTPCodeELB503CountAlarm9245609E": {
      "Properties": {
        "ActionsEnabled": false,
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "LoadBalancer",
            "Value": "default",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "HTTPCode_ELB_503_Count",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ELBOther5XXAlarmHTTPCodeELB504CountAlarm3CB4C2A2": {
      "Properties": {
        "ActionsEnabled": false,
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "LoadBalancer",
            "Value": "default",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "HTTPCode_ELB_504_Count",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ELBOther5XXAlarmHTTPCodeELB5XXCountAlarmC0D9FEA7": {
      "Properties": {
        "ActionsEnabled": false,
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "LoadBalancer",
            "Value": "default",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "HTTPCode_ELB_5XX_Count",
        "Namespace": "AWS/ApplicationELB",
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TopicBFC7AF6E": {
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
